miner/collator.go

```diff
@@ -0,0 +1,402 @@
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+// Package miner implements Ethereum block creation and mining.
+package miner
+
+import (
+	"errors"
+	//	"math"
+	//"math/big"
+	"os"
+	"plugin"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/core/vm"
+	//    "github.com/ethereum/go-ethereum/core/state"
+	"github.com/ethereum/go-ethereum/params"
+	"github.com/naoina/toml"
+)
+
+type CollatorAPI interface {
+	Version() string
+	Service() interface{}
+}
+
+// Pool is an interface to the transaction pool
+type Pool interface {
+	Pending(bool) (map[common.Address]types.Transactions, error)
+	Locals() []common.Address
+}
+
+/*
+	BlockState represents an under-construction block.  An instance of
+	BlockState is passed to CollateBlock where it can be filled with transactions
+	via BlockState.AddTransaction() and submitted for sealing via
+	BlockState.Commit().
+
+	Operations on a single BlockState instance are not threadsafe.  However,
+	instances can be copied with BlockState.Copy().
+*/
+type BlockState interface {
+	/*
+		adds a single transaction to the blockState.  Returned errors include ..,..,..
+		which signify that the transaction was invalid for the current EVM/chain state.
+
+		ErrRecommit signals that the recommit timer has elapsed.
+		ErrNewHead signals that the client has received a new canonical chain head.
+		All subsequent calls to AddTransaction fail if either newHead or the recommit timer
+		have occured.
+
+		If the recommit interval has elapsed, the BlockState can still be committed to the sealer.
+	*/
+	AddTransactions(tx types.Transactions) (error, types.Receipts)
+
+	/*
+		removes a number of transactions from the block resetting the state to what
+		it was before the transactions were added.  If count is greater than the number
+		of transactions in the block,  returns
+	*/
+	RevertTransactions(count uint) error
+
+	/*
+		returns true if the Block has been made the current sealing block.
+		returns false if the newHead interrupt has been triggered.
+		can also return false if the BlockState is no longer valid (the call to CollateBlock
+		which the original instance was passed has returned).
+	*/
+	Commit() bool
+	Copy() BlockState
+	State() vm.StateReader
+	Signer() types.Signer
+	Header() *types.Header
+	/*
+		the account which will receive the block reward.
+	*/
+	Etherbase() common.Address
+}
+
+const (
+	InterruptNone int = iota
+	InterruptResubmit
+	InterruptNewHead
+)
+
+/*
+InterruptContext allows for active polling to detect if a new canonical chain
+head was received or recommit timer elapse occured.
+*/
+type InterruptContext interface {
+	InterruptState() int
+}
```

Comment:
I don't think this is needed. Using a `context.Context` is a more natural thing in Go world. I do see that you have 2 different interrupt types, but I don't think it's important from the collator's perspective. It should either stop generating, or should not interrupt imho.

I guess you're trying to somehow have the outside miner direct retrying the same block with a new set of txs. That perhaps should be left to the collator? I.e. if the collator wants, it can keep running and remaking blocks every 3/10/whatnot seconds based on the potentially new contents of the txs. That would keep this feature collator specific (e.g. it doesn't make sense t resubmit with a mev code), and we could get rid of the API complexity.